include/experimental/__p2642_bits/layout_padded.hpp

```diff
@@ -0,0 +1,405 @@
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 4.0
+//       Copyright (2022) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+#pragma once
+
+#include <cassert>
+#include "../__p0009_bits/dynamic_extent.hpp"
+#include "../__p0009_bits/extents.hpp"
+#include "../__p0009_bits/mdspan.hpp"
+#include "../__p0009_bits/layout_left.hpp"
+#include "../__p0009_bits/layout_right.hpp"
+#include "../__p0009_bits/layout_stride.hpp"
+
+namespace std {
+namespace experimental {
+
+template <size_t padding_stride>
+struct layout_left_padded;
+
+template <size_t padding_stride>
+struct layout_right_padded;
+
+namespace detail {
+
+// offset_index_sequence idea comes from "offset_sequence" here:
+// https://devblogs.microsoft.com/oldnewthing/20200625-00/?p=103903
+//
+// offset_index_sequence adds N to each element of the given IndexSequence.
+// We can't just template on the parameter pack of indices directly;
+// the pack needs to be contained in some type.
+// We choose index_sequence because it stores no run-time data.
+template <size_t N, class IndexSequence>
+struct offset_index_sequence;
+
+template <size_t N, size_t... Indices>
+struct offset_index_sequence<N, index_sequence<Indices...>>
+{
+  using type = index_sequence<(Indices + N)...>;
+};
+
+template <size_t N, typename IndexSequence>
+using offset_index_sequence_t = typename offset_index_sequence<N, IndexSequence>::type;
+
+// iota_index_sequence defines the half-open sequence
+// begin, begin+1, begin+2, ..., end-1.
+// If end == begin, then the sequence is empty (we permit this).
+//
+// Defining the struct first, rather than going straight to the type alias,
+// lets us check the template arguments.
+template <size_t begin, size_t end>
+struct iota_index_sequence
+{
+  static_assert(end >= begin, "end must be >= begin.");
+  using type =
+      offset_index_sequence_t<begin, std::make_index_sequence<end - begin>>;
+};
+
+template <class _Extents, class _Enabled = void>
+struct __p_left;
```

Comment:
Oh hmm, I was just copying the convention used in mdspan since I know it was directly used in some library implementations. I'd be happy to switch it back to not using underscores because it is way more readable.

Reply:
we generally did this since it kinda std library. We should discuss our policy.